hoomd/custom_action.py

```diff
@@ -0,0 +1,81 @@
+from abc import ABC, abstractmethod
+from hoomd.parameterdicts import ParameterDict
+
+
+class _CustomAction(ABC):
+    """Base class for all Python ``Action``s.
+
+    This class must be the parent class for all Python ``Action``s. This class
+    requires all subclasses to implement the act method which performs the
+    Python object's task whether that be updating the system, writing output, or
+    analyzing some property of the system.
+
+    To use subclasses of this class, the object must be passed as an argument
+    for the :py:class:`hoomd.python_action._CustomOperation` constructor.
+
+    If the pressure, rotational kinetic energy, or external field virial is
+    needed for a subclass, the flags attribute of the class needs to be set with
+    the appropriate flags from :py:class:`hoomd.util.ParticleDataFlags`.
+
+    For advertising loggable quantities through the
+    :py:class:`hoomd.python_action._CustomOperation` object, the class attribute
+    ``log_quantities`` can be used. The dictionary expects string keys with the
+    name of the loggable and :py:class:`LoggerQuantity` objects as the values.
+    """
+    flags = []
+    log_quantities = {}
+
+    def __init__(self):
+        pass
+
+    def attach(self, simulation):
+        self._state = simulation.state
+
+    def detach(self):
+        if hasattr(self, '_state'):
+            del self._state
```

Comment:
As long as there are live references to the state, the garbage collector shouldn't collect it correct?

Reply:
For some reason when reviewing this I had an idea of how this might break the usual reference counting pattern, but looking again I can't think of any reason. You are right, it should be fine. I think it's more natural to set this to `None` (also in the constructor): even though working flexibly with attributes is a strength of Python using `hasattr` here still feels a bit more hackish. However, this solution will work fine.

Reply:
The reason, I don't do that is to limit references to the state and simulation. That is what we do internally at least (also because there is no universal `__init__` with standard `Operation`s); however, I can see an argument for something different with custom ones.

Reply:
I'm not sure I understand what you mean. How would setting it to `None` vs deleting the attribute be any different from the perspective of the reference counter?

Reply:
Well it is more just I didn't want to store references to the `_simulation` unless necessary, so many objects simply did not have the attribute. However, with #642 this may be a moot point.